dashboard/app/models/script.rb

```diff
@@ -955,143 +945,28 @@ def self.add_script(options, raw_lesson_groups, new_suffix: nil, editor_experime
         raw_lessons << temp_lesson
       end
     end
-
     raw_script_levels = raw_lessons.map {|lesson| lesson[:script_levels]}.flatten
-    script = fetch_script(options)
-    script.update!(hidden: true) if new_suffix
-    chapter = 0
-    lesson_position = 0; script_level_position = Hash.new(0)
-    script_lessons = []
-    script_lesson_groups = []
-    script_levels_by_lesson = {}
-    levels_by_key = script.levels.index_by(&:key)
-    lockable_count = 0
-    non_lockable_count = 0
-
-    unless raw_script_levels.empty?
-      # We want a script to either have all of its lessons have lesson groups
-      # or none of the lessons have lesson groups. We know we have hit this case if
-      # there are more than one lesson group for a script but the lesson group key
-      # for the first lesson is blank
-      if raw_lesson_groups[0][:key].nil? && raw_lesson_groups.length > 1
-        raise "Expect if one lesson has a lesson group all lessons have lesson groups. Lesson #{raw_lesson_groups[0][:lessons][0][:name]} does not have a lesson group."
-      elsif raw_lesson_groups.length == 1 && raw_lesson_groups[0][:key].nil?
-        lesson_group = LessonGroup.find_or_create_by(
-          key: '',
-          script: script,
-          user_facing: false,
-          position: 1
-        )
 
-        script_lesson_groups << lesson_group
-      else
-        # Finds or creates Lesson Groups with the correct position.
-        # In addition it check for 3 things:
-        # 1. That all the lesson groups specified by the editor have a key and
-        # display name.
-        # 2. If the lesson group key is an existing key that the given display name
-        # for that key matches the already saved display name
-        # 3. PLC courses use certain lesson group keys for module types. We reserve those
-        # keys so they can only map to the display_name for their PLC purpose
-        raw_lesson_groups&.each_with_index do |raw_lesson_group, index|
-          Plc::LearningModule::RESERVED_LESSON_GROUPS_FOR_PLC.each do |reserved_lesson_group|
-            if reserved_lesson_group[:key] == raw_lesson_group[:key] && reserved_lesson_group[:display_name] != raw_lesson_group[:display_name]
-              raise "The key #{reserved_lesson_group[:key]} is a reserved key. It must have the display name: #{reserved_lesson_group[:display_name]}."
-            end
-          end
-          if raw_lesson_group[:display_name].blank?
-            raise "Expect all lesson groups to have display names. The following lesson group does not have a display name: #{raw_lesson_group[:key]}"
-          end
+    script.lesson_groups = raw_script_levels.empty? ? [] : LessonGroup.add_lesson_groups(raw_lesson_groups, script)
```

Comment:
Why does the check need to be on `raw_script_levels` - would checking on `raw_lesson_groups` also work?

Ideally IMO, this conditional logic would be moved inside of `add_lesson_groups` if possible.

Reply:
I remove it in a later PR. Was just a hold over from the old format and was trying to move away from it slowly.